pygmt/clib/session.py

```diff
@@ -1792,43 +1791,43 @@ def virtualfile_in(  # noqa: PLR0912
             "vectors": self.virtualfile_from_vectors,
         }[kind]
 
-        # Ensure the data is an iterable (Python list or tuple)
-        if kind in {"geojson", "grid", "image", "file", "arg", "stringio"}:
-            if kind == "image" and data.dtype != "uint8":
-                msg = (
-                    f"Input image has dtype: {data.dtype} which is unsupported, "
-                    "and may result in an incorrect output. Please recast image "
-                    "to a uint8 dtype and/or scale to 0-255 range, e.g. "
-                    "using a histogram equalization function like "
-                    "skimage.exposure.equalize_hist."
-                )
-                warnings.warn(message=msg, category=RuntimeWarning, stacklevel=2)
-            _data = (data,) if not isinstance(data, pathlib.PurePath) else (str(data),)
-        elif kind == "empty":
-            # data is None, so data must be given via x/y/z.
-            _data = [x, y]
-            if z is not None:
-                _data.append(z)
-            if extra_arrays:
-                _data.extend(extra_arrays)
-        elif kind == "vectors":
-            if hasattr(data, "items") and not hasattr(data, "to_frame"):
-                # pandas.DataFrame or xarray.Dataset types.
-                # pandas.Series will be handled below like a 1-D numpy.ndarray.
-                _data = [array for _, array in data.items()]
-            else:
-                # Python list, tuple, numpy.ndarray, and pandas.Series types
-                _data = np.atleast_2d(np.asanyarray(data).T)
-        elif kind == "matrix":
-            # GMT can only accept a 2-D matrix which are signed integer (i), unsigned
-            # integer (u) or floating point (f) types. For other data types, we need to
-            # use virtualfile_from_vectors instead, which turns the matrix into a list
-            # of vectors and allows for better handling of non-integer/float type inputs
-            # (e.g. for string or datetime data types).
-            _data = (data,)
-            if data.dtype.kind not in "iuf":
-                _virtualfile_from = self.virtualfile_from_vectors
-                _data = data.T
+        # Ensure the data is an iterable (Python list or tuple).
+        match kind:
+            case "arg" | "file" | "geojson" | "grid" | "image" | "stringio":
+                _data = (data,)
```

Comment:
`_data = (data,) if not isinstance(data, pathlib.PurePath) else (str(data),)` is changed to `_data = (data,)`. 

For reference, the code `str(data)` was initially added in https://github.com/GenericMappingTools/pygmt/pull/1382 to support passing pathlib.Path as filenames. It turns out it's no longer necessary because the `build_arg_list` function already handle with `pathlib.Path` objects correctly:

https://github.com/GenericMappingTools/pygmt/blob/d41922ca4b4122e031d65d82bddea33798ea2ffd/pygmt/helpers/utils.py#L516-L520.

The support of pathlib.Path objects can even be tracked back to the old, deprecated `build_arg_string` function (xref: https://github.com/GenericMappingTools/pygmt/pull/1837).

We already have a lot of tests that takes a pathlib.Path object as input (e.g., https://github.com/GenericMappingTools/pygmt/blob/d41922ca4b4122e031d65d82bddea33798ea2ffd/pygmt/tests/test_info.py#L20). So the changes should be safe if all tests pass.